### PR TITLE
fix conv2d and conv3d support asymmetric padding and channel last

### PR DESCRIPTION
**Pull request by liym27**
1.support asymmetric padding;
    2.support padding algorithm:"SAME" and "VALID";
    3.support channel_last: data_format NHWC and NDHWC;
    4.change doc of python API and c++;
1.支持非对称padding;
2.支持padding 算法："SAME"和 "VALID";
3.支持channel last格式（NHWC NDHWC）;

中文文档修改PR：https://github.com/PaddlePaddle/FluidDoc/pull/1441

### CODE REVIEW

**Review by lanxianghit — DISMISSED** [review]

**Review by heavengate — DISMISSED** [review]
LG API, please ask @chengduoZH or @qingqing01 to check code correctness

File: paddle/fluid/API.spec [hunk 1]
```diff
@@ -129,8 +129,8 @@ paddle.fluid.layers.bpr_loss (ArgSpec(args=['input', 'label', 'name'], varargs=N
...
-paddle.fluid.layers.conv2d (ArgSpec(args=['input', 'num_filters', 'filter_size', 'stride', 'padding', 'dilation', 'groups', 'param_attr', 'bias_attr', 'use_cudnn', 'act', 'name'], varargs=None, keywords=None, defaults=(1, 0, 1, None, None, None, True, None, None)), ('document', '06de9adb5994f6f8cb806c75b55550af'))
-paddle.fluid.layers.conv3d (ArgSpec(args=['input', 'num_filters', 'filter_size', 'stride', 'padding', 'dilation', 'groups', 'param_attr', 'bias_attr', 'use_cudnn', 'act', 'name'], varargs=None, keywords=None, defaults=(1, 0, 1, None, None, None, True, None, None)), ('document', '71b09227709475fa178c1739dff64af6'))
+paddle.fluid.layers.conv2d (ArgSpec(args=['input', 'num_filters', 'filter_size', 'stride', 'padding', 'dilation', 'groups', 'param_attr', 'bias_attr', 'use_cudnn', 'act', 'name', 'data_format'], varargs=None, keywords=None, defaults=(1, 0, 1, None, None, None, True, None, None, 'NCHW')), ('document', 'b8da17862ba02b5297a37d2edd571d76'))
```
Comment (gongweibao) [inline paddle/fluid/API.spec:142]: `data_format` or `data_layout`?

File: paddle/fluid/operators/conv_cudnn_op.cu [hunk 2]
```diff
@@ -0,0 +1,1060 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (gongweibao) [inline paddle/fluid/operators/conv_cudnn_op.cu:1]: Why change from conv_cudnn_op.cc => conv_cudnn_op.cu?

File: paddle/fluid/operators/conv_op.cc [hunk 3]
```diff
@@ -167,17 +192,24 @@ void Conv2DOpMaker::Make() {
       .AsDispensable();
   AddOutput("Output",
             "(Tensor) The output tensor of convolution operator. "
-            "The format of output tensor is also NCHW.");
+            "The format of output tensor is also NCHW or NHWC.");
```
Comment (gongweibao) [inline paddle/fluid/operators/conv_op.cc]: It has same data layout and data type with the input.

File: paddle/fluid/operators/conv_op.cc [hunk 4]
```diff
@@ -620,7 +663,8 @@ framework::OpKernelType ConvOpDoubleGrad::GetExpectedKernelType(
   int customized_type_value =
       framework::OpKernelType::kDefaultCustomizedTypeValue;
   framework::LibraryType library_{framework::LibraryType::kPlain};
-  std::string data_format = ctx.Attr<std::string>("data_format");
+  // std::string data_format = ctx.Attr<std::string>("data_format");
```
Comment (gongweibao) [inline paddle/fluid/operators/conv_op.cc]: Clean up this?

File: paddle/fluid/API.spec (hunk 1 above)
Comment (liym27) [reply paddle/fluid/API.spec:142]: data_format. 已跟lanxianghit讨论过，目前统一用data_format。

File: paddle/fluid/operators/conv_op.cc (hunk 3 above)
Comment (liym27) [reply paddle/fluid/operators/conv_op.cc]: done

File: paddle/fluid/operators/conv_op.cc (hunk 4 above)
Comment (liym27) [reply paddle/fluid/operators/conv_op.cc]: done

File: paddle/fluid/operators/conv_cudnn_op.cu (hunk 2 above)
Comment (liym27) [reply paddle/fluid/operators/conv_cudnn_op.cu:1]: conv_cudnn_op.cu.cc  caused compilation failure.

**Review by raindrops2sea — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

File: paddle/fluid/operators/conv_cudnn_op.cu (hunk 2 above)
Comment (luotao1) [reply paddle/fluid/operators/conv_cudnn_op.cu:1]: 如果是编译错误的话，需要解决。

File: cmake/operators.cmake [hunk 5]
```diff
@@ -9,6 +9,7 @@ function(op_library TARGET)
     set(miopen_hip_cc_srcs)
     set(cu_cc_srcs)
     set(cudnn_cu_cc_srcs)
+    set(cudnn_cu_srcs)
     set(CUDNN_FILE)
     set(mkldnn_cc_srcs)
     set(MKLDNN_FILE)
```
Comment (luotao1) [inline cmake/operators.cmake:15]: 为什么还要改cmake呢？具体是什么编译错误？

**Comment by luotao1** [comment]
请 @qingqing01 再看下为什么从 conv_cudnn_op.cc => conv_cudnn_op.cu，这样改法是否可以？

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Comment by liym27** [comment]
> 请 @qingqing01 再看下为什么从 conv_cudnn_op.cc => conv_cudnn_op.cu，这样改法是否可以？

已确认可以。

File: cmake/operators.cmake (hunk 5 above)
Comment (liym27) [reply cmake/operators.cmake:15]: 改cmake是因为需要在pybind中成功加入 USE_OP_DEVICE_KERNEL(conv2d, CUDNN)

File: paddle/fluid/API.spec (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/API.spec:142]: https://github.com/PaddlePaddle/Paddle/pull/19614/files

It uses data_layout. Do we need to unify them?

File: paddle/fluid/operators/conv_cudnn_op.cu [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+    paddings[i].second = pads[i * 2 + 1];
+  }
+
+  auto src_tensor = EigenTensor<T, D>::From(src);
+  auto out_tensor = EigenTensor<T, D>::From(*out);
+
+  auto& place =
+      *context.template device_context<DeviceContext>().eigen_device();
+  out_tensor.device(place) = src_tensor.pad(paddings, pad_value);
+}
```
Comment (chengduoZH) [inline paddle/fluid/operators/conv_cudnn_op.cu:98]: Why don't you call the `PadFunction ` of padding.h but copying this function to here???
https://github.com/PaddlePaddle/Paddle/blob/1a3eef026c1d07516fcfd3ea33add8b3e1de169f/paddle/fluid/operators/math/padding.h#L28-L45

**Review by chengduoZH — COMMENTED** [review]
**You should move the padding logic to another file or reuse the logic of padding.h.**

File: paddle/fluid/operators/conv_cudnn_op.cu (hunk 6 above)
Comment (liym27) [reply paddle/fluid/operators/conv_cudnn_op.cu:98]: Modidied.

**Comment by liym27** [comment]
> **You should move the padding logic to another file or reuse the logic of padding.h.**

modified. Thanks

File: paddle/fluid/API.spec (hunk 1 above)
Comment (liym27) [reply paddle/fluid/API.spec:142]: Yes.  Some OPs use "data_layout" and others use "data_format". We will unify them layter.

**Pull request merged**